### PR TITLE
Convert thrust::optional usages to std::optional

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
As noted in https://github.com/rapidsai/cudf/issues/11368 we should strive towards not having thrust types in our 'public' API. 
This removes occurences of using `thrust::optional` from cudf/io host classes in preference of `std::optional`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/io/types.hpp [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (bdice) [inline cpp/include/cudf/io/types.hpp:1]:
```suggestion
```

**Review by bdice — COMMENTED** [review]
(edit:) Should we only limit this change to host code? (Can device code use `std::optional`?)

**Review by ttnghia — APPROVED** [review]

**Review by upsj — APPROVED** [review]

**Review by bdice — APPROVED** [review]

**Comment by robertmaynard** [comment]
> (edit:) Should we only limit this change to host code? (Can device code use `std::optional`?)

Only host code can use std::optional

**Comment by robertmaynard** [comment]
@gpucibot merge

**Pull request merged**